tachyon/math/polynomials/univariate/lagrange_interpolation.h

```diff
@@ -0,0 +1,68 @@
+#ifndef TACHYON_MATH_POLYNOMIALS_UNIVARIATE_LAGRANGE_INTERPOLATION_H_
+#define TACHYON_MATH_POLYNOMIALS_UNIVARIATE_LAGRANGE_INTERPOLATION_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/math/polynomials/univariate/univariate_polynomial.h"
+
+namespace tachyon::math {
+
+template <typename F, size_t MaxDegree>
+UnivariateDensePolynomial<F, MaxDegree> LagrangeInterpolate(
+    const std::vector<F>& points, const std::vector<F>& evals) {
+  using Poly = UnivariateDensePolynomial<F, MaxDegree>;
+  using Coeffs = UnivariateDenseCoefficients<F, MaxDegree>;
+
+  CHECK_EQ(points.size(), evals.size());
+  CHECK_EQ(points.size(), MaxDegree + 1);
+
+  if (points.size() == 1) {
+    return Poly(Coeffs({evals[0]}));
+  }
+
+  // points = [x₀, x₁, ..., xₙ]
+  // denoms[j][k] = (xⱼ - xₖ)
+  std::vector<std::vector<F>> denoms(points.size());
+  for (size_t j = 0; j < points.size(); ++j) {
+    for (size_t k = 0; k < points.size(); ++k) {
+      if (k != j) {
+        denoms[j].push_back(points[j] - points[k]);
+      }
+    }
+  }
+
+  // denoms[i][j] = 1 / (xⱼ - xₖ)
+  for (auto& denom : denoms) {
+    F::BatchInverseInPlace(denom);
+  }
+
+  std::vector<F> final_coeffs(points.size(), F::Zero());
```

Comment:
You are missing comments from here, and i believe this can be parallelized, too.